File: compilers/go/compiler/n133_loop_start_test.go
package compiler

import (
	"fmt"
	"testing"
)

// ---------------------------------------------------------------------------
// N-133 — `loop.start` on the `--ir` trust boundary
// ---------------------------------------------------------------------------
//
// `Loop.start` is `integer | string`, and the string arm is the sanctioned
// `"<decimal>n"` form — the suffix is REQUIRED, which is what Java's loader
// implements and what the schema's other two string-carrying integer fields
// (`load_const.value`, `ANFProperty.initialValue`) mean by a string.
//
// Go's `decodeBigIntFromRaw` stripped the trailing `n` only IF one was there
// and then parsed the rest as decimal either way, so `"5"` read as 5 here —
// while Rust read the same input as 0. Two tiers, two different loops, both
// exit 0. The suffix is the only thing that makes the string arm
// unambiguous, so Go requires it.
//
// The probe is a two-iteration loop summing its iterator, so the start lands
// in the emitted bytes and nothing else does.

func n133IR(startJSON string) []byte {
	return []byte(fmt.Sprintf(`{
	  "contractName": "LoopProbe",
	  "properties": [{"name": "target", "type": "bigint", "readonly": true}],
	  "methods": [
	    {"name": "constructor", "params": [], "isPublic": false,
	     "body": [{"name": "t0", "value": {"kind": "call", "func": "super", "args": []}}]},
	    {"name": "run", "params": [], "isPublic": true,
	     "body": [
	       {"name": "acc", "value": {"kind": "load_const", "value": 0}},
	       {"name": "t1", "value": {"kind": "loop", "count": 2, "iterVar": "i",
	         "start": %s, "step": 1,
	         "body": [{"name": "acc", "value": {"kind": "bin_op", "left": "acc", "op": "+", "right": "i"}}]}},
	       {"name": "t2", "value": {"kind": "load_prop", "name": "target"}},
	       {"name": "t3", "value": {"kind": "bin_op", "left": "acc", "op": "===", "right": "t2"}},
	       {"name": "t4", "value": {"kind": "assert", "value": "t3"}}
	     ]}
	  ]
	}`, startJSON))
}

func n133Hex(t *testing.T, startJSON string) string {
	t.Helper()
	art, err := CompileFromIRBytes(n133IR(startJSON))
	if err != nil {
		t.Fatalf("loop.start %s: unexpected error: %v", startJSON, err)
	}
	return art.Script
}

// The sanctioned form must lower to the SAME BYTES as the integer it spells.
// Both cases non-zero: 0 is what every fallback path also produces, and it is
// what the loop this probe is modelled on actually carries.
func TestN133_DecimalBigIntStringStart_MeansTheInteger(t *testing.T) {
	for _, c := range []struct{ str, integer, want string }{
		{`"5n"`, `5`, "5b009c"},
		{`"-3n"`, `-3`, "0185009c"},
	} {
		gotStr := n133Hex(t, c.str)
		gotInt := n133Hex(t, c.integer)
		if gotStr != gotInt {
			t.Errorf("start %s produced %s, integer %s produced %s",
				c.str, gotStr, c.integer, gotInt)
		}
		if gotStr != c.want {
			t.Errorf("start %s: got %s, want %s", c.str, gotStr, c.want)
		}
	}
}

// A start Go can represent but that is wider than any tier's native int64.
// Go carries `start` as a *big.Int, so it emits the value.
func TestN133_OverInt64Start_IsNotTruncated(t *testing.T) {
	const want = "0dffffff7fd4dbe98ca039593e19009c"
	if got := n133Hex(t, `"999999999999999999999999999999n"`); got != want {
		t.Errorf("over-int64 start: got %s, want %s", got, want)
	}
}

// Anything that is neither a JSON integer nor the `"<decimal>n"` form is
// refused. `"5"` is the row this change adds: it used to read as 5 here and
// as 0 in Rust.
func TestN133_UnreadableStart_IsRefused(t *testing.T) {
	for _, bad := range []string{`"5"`, `"abc"`, `""`, `"5nn"`, `"n"`, `"-n"`, `"1.5n"`, `true`, `null`} {
		if _, err := CompileFromIRBytes(n133IR(bad)); err == nil {
			t.Errorf("loop.start %s: expected an error, got none", bad)
		}
	}
}

// The start-0 script, so a future fallback has something to collide with:
// every refusal above must NOT be reachable as "exit 0 with these bytes".
func TestN133_ZeroStart_IsItsOwnScript(t *testing.T) {
	if got := n133Hex(t, `0`); got != "008b009c" {
		t.Errorf("start 0: got %s, want 008b009c", got)
	}
	if n133Hex(t, `0`) == n133Hex(t, `5`) {
		t.Error("probe is vacuous: start 0 and start 5 emit the same bytes")
	}
}

File: compilers/zig/src/ir/n133_loop_start_test.zig
//! N-133 — an unreadable `loop.start` must produce an ERROR, never 0.
//!
//! `Loop.start` is `integer | string` and the string arm is the sanctioned
//! `"<decimal>n"` form (issue #121). This tier read it as:
//!
//!     .string => |s| blk: {
//!         const text = if (s[s.len - 1] == 'n') s[0 .. s.len - 1] else s;
//!         break :blk std.fmt.parseInt(i64, text, 10) catch 0;   // <- here
//!     },
//!     else => 0,                                                // <- and here
//!
//! Both fallbacks answer 0, and 0 is the worst possible answer: it is a
//! perfectly plausible loop start — the commonest one, and the one
//! `bounded-loop`'s own golden carries — so the wrong program compiles, emits
//! a well-formed locking script, and nothing anywhere looks wrong.
//!
//! Measured on that golden with only `start` edited, fold-off:
//!
//!     "999999999999999999999999999999n"
//!                  go/rust/python/ruby/java the value · ZIG the start-0 script
//!     "abc" / "" / "5nn" / true
//!                  go/python/java refuse    · rust/ZIG/ruby the start-0 script
//!
//! The `catch 0` case is not hypothetical. A source contract whose loop starts
//! past int64 compiles, and ts / go / rust write that start as
//! `"999999999999999999999999999999n"` in their `--emit-ir` output — which is
//! exactly the string this tier turned into 0.
//!
//! THE OVER-INT64 CASE IS A REAL SPLIT, and refusing is this tier's half of
//! it rather than a fix for it. `types.ANFLoop.start` is an `i64`, so 10^30 is
//! a value this tier genuinely cannot carry, while its five peers hold `start`
//! as an arbitrary-precision integer and emit the right script. Widening it is
//! separate work — `start` is an `i64` through the whole codegen path, not
//! just at the loader. What this change settles is only that an unrepresentable
//! start is REFUSED instead of silently replaced, which was a defect under
//! either reading.
//!
//! Every assertion here is one the pre-fix loader could not have passed by
//! accident: `expectError` cannot be satisfied by a loader that returns a
//! program, and `catch 0` always returned one.

const std = @import("std");
const testing = std.testing;
const json = @import("json.zig");

/// Smallest IR carrying one `loop` whose `start` is the spliced-in text.
fn irWithStart(allocator: std.mem.Allocator, start: []const u8) ![]const u8 {
    return std.fmt.allocPrint(allocator,
        \\{{"contractName":"Probe","properties":[],"methods":[{{"name":"unlock","params":[],"isPublic":true,"body":[{{"name":"t0","value":{{"kind":"loop","count":2,"start":{s},"step":1,"iterVar":"i","body":[{{"name":"t1","value":{{"kind":"load_const","value":0}}}}]}}}}]}}]}}
    , .{start});
}

fn expectStartRefused(a: std.mem.Allocator, start: []const u8) !void {
    const src = try irWithStart(a, start);
    try testing.expectError(error.InvalidLoopStart, json.parseANFProgram(a, src));
}

fn expectStartLoads(a: std.mem.Allocator, start: []const u8, want: i64) !void {
    const src = try irWithStart(a, start);
    const program = try json.parseANFProgram(a, src);
    const value = program.methods[0].body[0].value;
    try testing.expectEqual(want, value.loop.start);
}

test "N-133: a loop start string without the `n` suffix is refused" {
    // The row that shows how bad the substitution is: go, python, zig and ruby
    // all read "5" as 5 while rust read the SAME input as 0, and both sides
    // exited 0 with a well-formed script. There is no majority answer to
    // adopt, so the sanctioned form is the rule and "5" is not it.
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    try expectStartRefused(arena.allocator(), "\"5\"");
}

test "N-133: a non-numeric loop start string is refused, not read as 0" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const a = arena.allocator();
    try expectStartRefused(a, "\"abc\"");
    try expectStartRefused(a, "\"\"");
    try expectStartRefused(a, "\"n\"");
    try expectStartRefused(a, "\"-n\"");
}

test "N-133: exactly one trailing `n` is stripped, so \"5nn\" stays refused" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    try expectStartRefused(arena.allocator(), "\"5nn\"");
}

test "N-133: a float-shaped string cannot reopen the float boundary" {
    // N-131 closed the numeric float arm. `"1.5n"` is the way back in through
    // the string arm, and `parseInt` refusing it used to mean 0, not an error.
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    try expectStartRefused(arena.allocator(), "\"1.5n\"");
}

test "N-133: a non-integer, non-string loop start is refused" {
    // The `else => 0` arm: a boolean, a null, an object, an array.
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const a = arena.allocator();
    try expectStartRefused(a, "true");
    try expectStartRefused(a, "null");
    try expectStartRefused(a, "{}");
    try expectStartRefused(a, "[]");
}

test "N-133: a start wider than i64 is refused rather than replaced by 0" {
    // The documented split. This tier cannot represent it; five peers can and
    // do. Refusing is the half that stops a DIFFERENT loop being compiled.
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const a = arena.allocator();
    try expectStartRefused(a, "\"999999999999999999999999999999n\"");
    // A bare JSON number past i64 lands in std.json's `number_string` arm,
    // which the same `else => 0` used to swallow. python, ruby and java all
    // WRITE the over-int64 start in exactly this shape.
    try expectStartRefused(a, "999999999999999999999999999999");
    // The boundary itself: i64 max loads, i64 max + 1 does not.
    try expectStartLoads(a, "9223372036854775807", 9223372036854775807);
    try expectStartRefused(a, "9223372036854775808");
    try expectStartLoads(a, "-9223372036854775808", -9223372036854775808);
    try expectStartRefused(a, "-9223372036854775809");
}

test "N-133 control: the sanctioned forms still load, as the numbers they spell" {
    // The teeth against an over-broad guard: a loader that refuses the bad
    // starts by refusing strings, or by refusing loops, reddens here.
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const a = arena.allocator();
    try expectStartLoads(a, "\"5n\"", 5);
    try expectStartLoads(a, "\"-3n\"", -3);
    try expectStartLoads(a, "\"0n\"", 0);
    try expectStartLoads(a, "5", 5);
    try expectStartLoads(a, "-3", -3);
    try expectStartLoads(a, "0", 0);
}

test "N-133 control: a loop with no start at all is still a zero-start loop" {
    // Older payloads omit `start` entirely. Absent is not the same as
    // unreadable, and must keep defaulting to 0 rather than becoming an error.
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const a = arena.allocator();
    const src =
        \\{"contractName":"Probe","properties":[],"methods":[{"name":"unlock","params":[],"isPublic":true,"body":[{"name":"t0","value":{"kind":"loop","count":2,"iterVar":"i","body":[{"name":"t1","value":{"kind":"load_const","value":0}}]}}]}]}
    ;
    const program = try json.parseANFProgram(a, src);
    try testing.expectEqual(@as(i64, 0), program.methods[0].body[0].value.loop.start);
}
